nav2_costmap_2d/plugins/plugin_container_layer.cpp

```diff
@@ -0,0 +1,235 @@
+}
+
+void PluginContainerLayer::updateBounds(
+  double robot_x,
+  double robot_y,
+  double robot_yaw,
+  double * min_x,
+  double * min_y,
+  double * max_x,
+  double * max_y)
+{
+  for (vector<std::shared_ptr<Layer>>::iterator plugin = plugins_.begin(); plugin != plugins_.end();
+    ++plugin)
+  {
+    (*plugin)->updateBounds(robot_x, robot_y, robot_yaw, min_x, min_y, max_x, max_y);
+  }
+}
+
+void PluginContainerLayer::updateCosts(
+  nav2_costmap_2d::Costmap2D & master_grid,
+  int min_i,
+  int min_j,
+  int max_i,
+  int max_j)
+{
+  std::lock_guard<Costmap2D::mutex_t> guard(*getMutex());
+  if (!enabled_) {
+    return;
+  }
+
+  resetMap(min_i, min_j, max_i, max_j);
+
+  for (vector<std::shared_ptr<Layer>>::iterator plugin = plugins_.begin(); plugin != plugins_.end();
+    ++plugin)
+  {
+    (*plugin)->updateCosts(*this, min_i, min_j, max_i, max_j);
+  }
+
+  switch (combination_method_) {
+    case CombinationMethod::Overwrite:
+      updateWithOverwrite(master_grid, min_i, min_j, max_i, max_j);
+      break;
+    case CombinationMethod::Max:
+      updateWithMax(master_grid, min_i, min_j, max_i, max_j);
+      break;
+    case CombinationMethod::MaxWithoutUnknownOverwrite:
+      updateWithMaxWithoutUnknownOverwrite(master_grid, min_i, min_j, max_i, max_j);
+      break;
+    default:  // Nothing
+      break;
+  }
+
+  current_ = true;
+}
+
+void PluginContainerLayer::activate()
+{
+  for (vector<std::shared_ptr<Layer>>::iterator plugin = plugins_.begin(); plugin != plugins_.end();
+    ++plugin)
+  {
+    (*plugin)->activate();
+  }
+}
+
+void PluginContainerLayer::deactivate()
+{
+  for (vector<std::shared_ptr<Layer>>::iterator plugin = plugins_.begin(); plugin != plugins_.end();
+    ++plugin)
+  {
+    (*plugin)->deactivate();
+  }
+}
+
+void PluginContainerLayer::reset()
+{
+  for (vector<std::shared_ptr<Layer>>::iterator plugin = plugins_.begin(); plugin != plugins_.end();
+    ++plugin)
+  {
+    (*plugin)->reset();
+  }
+
+  current_ = false;
+}
+
+void PluginContainerLayer::onFootprintChanged()
+{
+  for (vector<std::shared_ptr<Layer>>::iterator plugin = plugins_.begin(); plugin != plugins_.end();
+    ++plugin)
+  {
+    (*plugin)->onFootprintChanged();
+  }
+}
+
+void PluginContainerLayer::matchSize()
+{
+  std::lock_guard<Costmap2D::mutex_t> guard(*getMutex());
+  Costmap2D * master = layered_costmap_->getCostmap();
+  resizeMap(
+    master->getSizeInCellsX(), master->getSizeInCellsY(),
+    master->getResolution(), master->getOriginX(), master->getOriginY());
+
+  for (vector<std::shared_ptr<Layer>>::iterator plugin = plugins_.begin(); plugin != plugins_.end();
+    ++plugin)
+  {
+    (*plugin)->matchSize();
+  }
+}
+
+bool PluginContainerLayer::isClearable()
+{
+  for (vector<std::shared_ptr<Layer>>::iterator plugin = plugins_.begin(); plugin != plugins_.end();
+    ++plugin)
+  {
+    if(!(*plugin)->isClearable()) {
+      return false;
```

Comment:
I think this logic needs o be inverted. If any plugin is clearable, then return true. Else, false. 

Right now, if any one thing isn't clearable (i.e. inflation layer) then we don't clear any of it

Reply:
Got it, just to clarify, in a situation where one of the layers is clearable and the other ones are not, we'll just go ahead and return that this plugin is clearable.

Are there any safety implications in that scenario? Such as clearing sub layer that is not supposed to be cleared

Reply:
We use this here: https://github.com/ros-navigation/navigation2/blob/add09f60b17f6134416122f8285eba7268d275d7/nav2_costmap_2d/src/clear_costmap_service.cpp#L116-L119

When someone asks for a costmap clear, this is what tells us if we can / should do it. Your implementation would only clear this container's cost-grid, but wouldn't clear its constituent plugin costmap representations 

What we should do is implement the `clearArea` method (ex. https://github.com/SteveMacenski/spatio_temporal_voxel_layer/blob/a46e4bb47d175872a2681360901a47cd0b014f7c/spatio_temporal_voxel_layer/src/spatio_temporal_voxel_layer.cpp#L954-L968). Within that, clear the costmap datastructure of the layer & also clear out the region from any `isClearable` sub-layer

Reply:
swapped the logic